gson/src/test/java/com/google/gson/JsonArrayTest.java

```diff
@@ -35,6 +35,7 @@ public void testEqualsOnEmptyArray() {
   }
 
   @Test
+  @SuppressWarnings("TruthSelfEquals")
```

Comment:
This should not be suppressed, see [its description](https://errorprone.info/bugpattern/TruthSelfEquals).

`assertThat(a).isEqualTo(a);` below should be changed to `assertThat(a.equals(a)).isTrue()` (or `MoreAsserts.assertEqualsAndHashCode(a, a)`?) since this is intended to explicitly test the `equals` implementation.

Reply:
In #2299, Éamonn said we could use `com.google.guava:guava-testlib` to simplify assert like this. 

[Link to comment](https://github.com/google/gson/pull/2299#discussion_r1070699132)

I never use `guava-testlib` so I'm not sure how to do that. I have looked around for some documentations/examples, but I don't find nothing.

Reply:
Hmm, right. Personally I think for this PR it would be best to omit these suppressions and change the code to `a.equals(a)` to fix this issue. If desired we could then add `guava-testlib` in a separate PR.

Reply:
Yes, sorry, I should have been more explicit. That is exactly what you can write, and alternatively you can have just one test method with all the different kinds of objects in different equality groups. Something like this:

```java
new EqualsTester()
    .addEqualityGroup(array1, array2) // with equal contents
    .addEqualityGroup(array3, array4) // equal to each other but not array1 or array2
    .addEqualityGroup(object1, object2)
    ...
   .testEquals();
```

`EqualsTester` tests that each object in an equality group is equal to itself and to each other object in the group, and not equal to any object in any other group. It also checks that equal objects have equal hashcodes. So it removes a lot of drudgery.

I think that would be a fine project for a separate PR.